File: firebase/firestore/node/sections.dart
// @dart=2.9

import "package:firestore/data.dart";
import "package:firestore/faculty.dart";
import "package:firestore/helpers.dart";
import "package:firestore/sections.dart";
import "package:firestore/services.dart";

Future<void> main() async {
	Args.initLogger("Indexing data...");

	final Map<String, String> courseNames = await Logger.logValue(
		"course names", () async => SectionReader.courseNames
	);

	final Map<String, String> sectionTeachers = await Logger.logValue(
		"section teachers", SectionReader.getSectionFacultyIds
	);

	final Map<String, User> facultyNames = await Logger.logValue(
		"faculty names", FacultyReader.getFaculty,
	);

	final Map<String, String> zoomLinks = await Logger.logValue(
		"zoom links", ZoomLinkReader.getZoomLinks,
	);
	Logger.info("Found ${zoomLinks.keys.length} zoom links");

	final List<Section> sections = await Logger.logValue(
		"sections list", () => SectionLogic.getSections(
			courseNames: courseNames,
			sectionTeachers: sectionTeachers,
			facultyNames: facultyNames,
			zoomLinks: zoomLinks,
		)
	);

	Logger.info("Finished data indexing.");

	if (Args.upload) {
		await Logger.logProgress(
			"data upload", () => Firestore.uploadSections(sections)
		);
	} else {
		Logger.warning("Did not upload section data. Use the --upload flag.");
	}
	await app.delete();
	Logger.info("Processed ${sections.length} sections.");
}

File: firebase/firestore-py/bin/sections.py
from lib.utils import logger
from lib.sections import reader as section_reader
from lib.faculty import reader as faculty_reader
from lib.sections import logic as section_logic
from lib import utils
from lib import services

if __name__ == "__main__":
  print("Indexing data...")

  course_names = logger.log_value("course names", section_reader.get_course_names)

  section_teachers = logger.log_value("section teachers", section_reader.get_section_faculty_ids)

  faculty_names = logger.log_value("faculty names", faculty_reader.get_faculty)

  zoom_links = logger.log_value("zoom links", section_reader.get_zoom_links)

  print(f"Found {len(zoom_links.keys())} zoom links")

  sections = logger.log_value("sections list", lambda: section_logic.get_sections(
    course_names = course_names,
    section_teachers = section_teachers,
    faculty_names = faculty_names,
    zoom_links = zoom_links,
    )
  )

  print("Finished data indexing.")

  if utils.args.should_upload:
    utils.logger.log_progress(
      "data upload", lambda: services.upload_sections(sections)
    )
  else:
    print("Did not upload sections data. Use the --upload flag.")
  
  utils.logger.info(f"Processed {len(sections)} sections")

  
